File: TTTLive-OC/TTTLive/TTTAVRegion.m
//
//  TTTAVRegion.m
//  TTTLive
//
//  Created by yanzhen on 2018/8/21.
//  Copyright © 2018年 yanzhen. All rights reserved.
//

#import "TTTAVRegion.h"

@interface TTTAVRegion ()
@property (strong, nonatomic) IBOutlet UIView *backgroundView;
@property (weak, nonatomic) IBOutlet UIImageView *videoView;
@property (weak, nonatomic) IBOutlet UILabel *idLabel;
@property (weak, nonatomic) IBOutlet UILabel *audioStatsLabel;
@property (weak, nonatomic) IBOutlet UILabel *videoStatsLabel;
@property (weak, nonatomic) IBOutlet UIButton *voiceBtn;
@property (weak, nonatomic) IBOutlet UIButton *switchBtn;

@end

@implementation TTTAVRegion

- (instancetype)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        [[NSBundle mainBundle] loadNibNamed:@"TTTAVRegion" owner:self options:nil];
        _backgroundView.frame = self.bounds;
        _backgroundView.alpha = 0.7;
        [self addSubview:_backgroundView];
    }
    return self;
}

- (IBAction)enableAudioAction:(UIButton *)sender {
    if (TTManager.me != _user) { return; }
    sender.selected = !sender.isSelected;
    TTManager.me.mutedSelf = sender.isSelected;
    [TTManager.rtcEngine muteLocalAudioStream:sender.isSelected];
    [_voiceBtn setImage:[UIImage imageNamed:sender.isSelected ? @"audio_close" : @"audio_small"] forState:UIControlStateNormal];
}

- (IBAction)switchCamera:(id)sender {
    [TTManager.rtcEngine switchCamera];
}

- (TTTVideoPosition *)videoPosition {
    if (!_videoPosition) {
        _videoPosition = [[TTTVideoPosition alloc] init];
        CGRect convertFrame = [self.superview convertRect:self.frame toView:self.superview.superview.superview];
        CGFloat width = UIScreen.mainScreen.bounds.size.width;
        CGFloat height = UIScreen.mainScreen.bounds.size.height;
        _videoPosition.x = self.frame.origin.x / width;
        _videoPosition.y = convertFrame.origin.y / height;
        _videoPosition.w = self.frame.size.width / width;
        _videoPosition.h = self.frame.size.height / height;
    }
    return _videoPosition;
}
#pragma mark - public
- (void)configureRegion:(TTTUser *)user {
//    self.hidden = NO;
    self.user = user;
    _backgroundView.alpha = 1;
    [_voiceBtn setImage:[UIImage imageNamed:@"audio_small"] forState:UIControlStateNormal];
    _idLabel.hidden = NO;
    _voiceBtn.hidden = NO;
    _audioStatsLabel.hidden = NO;
    _videoStatsLabel.hidden = NO;
    _idLabel.text = [NSString stringWithFormat:@"%lld", user.uid];
    
    TTTRtcVideoCanvas *videoCanvas = [[TTTRtcVideoCanvas alloc] init];
    videoCanvas.uid = user.uid;
    videoCanvas.renderMode = TTTRtc_Render_Adaptive;
    videoCanvas.view = _videoView;
    if (TTManager.me == user) {
        [TTManager.rtcEngine setupLocalVideo:videoCanvas];
        _switchBtn.hidden = NO;
    } else {
        [TTManager.rtcEngine setupRemoteVideo:videoCanvas];
        if (user.mutedSelf) {
            [self mutedSelf:YES];
        }
    }
}

- (void)closeRegion {
//    self.hidden = YES;
    _backgroundView.alpha = 0.7;
    _idLabel.hidden = YES;
    _voiceBtn.hidden = YES;
    _switchBtn.hidden = YES;
    _audioStatsLabel.hidden = YES;
    _videoStatsLabel.hidden = YES;
    _videoView.image = [UIImage imageNamed:@"video_head"];
    _user = nil;
}

- (void)reportAudioLevel:(NSUInteger)level {
    if (_user.mutedSelf) { return; }
    [_voiceBtn setImage:[TTManager getVoiceImage:level] forState:UIControlStateNormal];
}

- (void)setLocalAudioStats:(NSUInteger)stats {
    _audioStatsLabel.text = [NSString stringWithFormat:@"A-↑%lukbps",(unsigned long)stats];
}

- (void)setLocalVideoStats:(NSUInteger)stats {
    _videoStatsLabel.text = [NSString stringWithFormat:@"V-↑%lukbps",(unsigned long)stats];
}

- (void)setRemoterAudioStats:(NSUInteger)stats {
    _audioStatsLabel.text = [NSString stringWithFormat:@"A-↓%lukbps",(unsigned long)stats];
}

- (void)setRemoterVideoStats:(NSUInteger)stats {
    _videoStatsLabel.text = [NSString stringWithFormat:@"V-↓%lukbps",(unsigned long)stats];
}

- (void)mutedSelf:(BOOL)mute {
    [_voiceBtn setImage:[UIImage imageNamed:mute ? @"muted" : @"audio_small"] forState:UIControlStateNormal];
}
@end

File: TTTLive/TTTLive/TTTAVRegion.swift
//
//  TTTAVRegion.swift
//  TTTLive
//
//  Created by yanzhen on 2018/8/16.
//  Copyright © 2018年 yanzhen. All rights reserved.
//

import UIKit
import TTTRtcEngineKit

class TTTAVRegion: UIView {

    public var user: TTTUser?
    @IBOutlet private var backgroundView: UIView!
    @IBOutlet private weak var videoView: UIImageView!
    @IBOutlet private weak var idLabel: UILabel!
    @IBOutlet private weak var audioStatsLabel: UILabel!
    @IBOutlet private weak var videoStatsLabel: UILabel!
    @IBOutlet private weak var voiceBtn: UIButton!
    
    @IBOutlet weak var switchBtn: UIButton!
    public lazy var videoPosition: TTTVideoPosition = {
        //Please After UI layout done
        let position = TTTVideoPosition()
        let screenW = UIScreen.main.bounds.size.width
        let screenH = UIScreen.main.bounds.size.height
        let convertFrame = superview?.convert(frame, to: superview?.superview?.superview)
        position.x = Double(frame.origin.x / screenW)
        position.y = Double(convertFrame!.origin.y / screenH)
        position.w = Double(bounds.size.width / screenW)
        position.h = Double(bounds.size.height / screenH)
        return position
    }()
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        loadNibView()
    }
    
    private func loadNibView() {
        let bundle = Bundle(for: type(of: self))
        let nib = UINib(nibName: "TTTAVRegion", bundle: bundle)
        backgroundView = nib.instantiate(withOwner: self, options: nil)[0] as! UIView
        backgroundView.frame = bounds
        backgroundView.alpha = 0.7
        backgroundView.backgroundColor = UIColor.clear
        addSubview(backgroundView)
    }
    
    @IBAction private func enableAudioAction(_ sender: UIButton) {
        if TTManager.me != user { return; }
        sender.isSelected = !sender.isSelected
        TTManager.me.mutedSelf = sender.isSelected
        TTManager.rtcEngine.muteLocalAudioStream(sender.isSelected)
        mutedSelf(TTManager.me.mutedSelf)
    }
    
    @IBAction private func switchBtnAction(_ sender: Any) {
        TTManager.rtcEngine.switchCamera()
    }
    
    private func getVoiceImage(_ audioLevel: UInt) -> UIImage {
        var image: UIImage = #imageLiteral(resourceName: "voice_small")
        if audioLevel < 4 {
            image = #imageLiteral(resourceName: "voice_small")
        } else if audioLevel < 7 {
            image = #imageLiteral(resourceName: "voice_middle")
        } else {
            image = #imageLiteral(resourceName: "voice_big")
        }
        return image
    }
}

extension TTTAVRegion {
    public func configureRegion(_ user: TTTUser) {
        self.user = user
        backgroundView.alpha = 1
        voiceBtn.setImage(#imageLiteral(resourceName: "voice_small"), for: .normal)
        idLabel.isHidden = false
        voiceBtn.isHidden = false
        audioStatsLabel.isHidden = false
        videoStatsLabel.isHidden = false
        idLabel.text = user.uid.description
        
        let videoCanvas = TTTRtcVideoCanvas()
        videoCanvas.uid = user.uid
        videoCanvas.renderMode = .render_Adaptive
        videoCanvas.view = videoView
        if user == TTManager.me {
            TTManager.rtcEngine.setupLocalVideo(videoCanvas)
            switchBtn.isHidden = false
        } else {
            switchBtn.isHidden = true
            TTManager.rtcEngine.setupRemoteVideo(videoCanvas)
            if user.mutedSelf {
                mutedSelf(true)
            }
        }
    }
    
    public func closeRegion() {
        idLabel.isHidden = true
        voiceBtn.isHidden = true
        switchBtn.isHidden = true
        audioStatsLabel.isHidden = true
        videoStatsLabel.isHidden = true
        backgroundView.alpha = 0.7
        user = nil
        videoView.image = #imageLiteral(resourceName: "video_head")
    }
    
    public func reportAudioLevel(_ audioLevel: UInt) {
        if user!.mutedSelf { return }
        voiceBtn.setImage(getVoiceImage(audioLevel), for: .normal)
    }
    
    public func setLocalAudioStats(_ stats: UInt) {
        audioStatsLabel.text = "A-↑\(stats)kbps"
    }
    
    public func setLocalVideoStats(_ stats: UInt) {
        videoStatsLabel.text = "V-↑\(stats)kbps"
    }
    
    public func setRemoterAudioStats(_ stats: UInt) {
        audioStatsLabel.text = "A-↓\(stats)kbps"
    }
    
    public func setRemoterVideoStats(_ stats: UInt) {
        videoStatsLabel.text = "V-↓\(stats)kbps"
    }
    
    public func mutedSelf(_ mute: Bool) {
        voiceBtn.setImage(mute ? #imageLiteral(resourceName: "speaking_closed") : #imageLiteral(resourceName: "voice_small"), for: .normal)
    }
}
